src/vec/vec-final.md

```diff
@@ -235,21 +235,22 @@ impl<T> Iterator for RawValIter<T> {
             None
         } else {
             unsafe {
-                let result = ptr::read(self.start);
-                self.start = if mem::size_of::<T>() == 0 {
-                    (self.start as usize + 1) as *const _
+                if mem::size_of::<T>() == 0 {
+                    self.start = (self.start as usize + 1) as *const _;
+                    Some(ptr::read(NonNull::<T>::dangling().as_ptr()))
                 } else {
-                    self.start.offset(1)
-                };
-                Some(result)
+                    let old_ptr = self.start;
+                    self.start = self.start.offset(1);
+                    Some(ptr::read(old_ptr))
+                }
```

Comment:
Would it make sense to factor out a single raw pointer that is `Some(ptr::read(elem_ptr))`'d *afterwards*?

Reply:
I think it's very "doesn't matter" because this is a static branch, but I am mildly more inclined to this impl because it's trivial to swap it out with the read_unaligned version if you want that instead

Reply:
Or I guess put another way: these are two wholy distinct implementations that we're shoveling into one function body, and trying to reuse code was the reason we had a bug in the first place, so let's just be more honest about that reality and not mix the code.